### PR TITLE
Runtime state set by Ignite (StartVM), ignite-spawn only handles shutdown

### PR DESCRIPTION
**Pull request by twelho**
With this `ignite-spawn` no longer writes to the VM metadata file on startup, it only performs cleanup on after VM shutdown. The running state, container ID, start time and VM IP addresses are now set by `StartVM` after starting the container (and waiting for `ignite-spawn`).

cc @luxas

### CODE REVIEW

**Comment by twelho** [comment]
This has an issue: `ignite-spawn` only manages to clear the state about 30% of the time. An unsuccessful clear leaves Ignite in a limbo, where the VM cannot be started or stopped anymore without editing the state by hand.

File: cmd/ignite-spawn/ignite-spawn.go [hunk 1]
```diff
@@ -47,7 +48,7 @@ func StartVM(vm *api.VM) error {
 	// Serve DHCP requests for those interfaces
 	// This function returns the available IP addresses that are being
 	// served over DHCP now
-	ipAddrs, err := container.StartDHCPServers(vm, dhcpIfaces)
+	_, err = container.StartDHCPServers(vm, dhcpIfaces)
```
Comment (luxas) [inline cmd/ignite-spawn/ignite-spawn.go]: Please cleanup the code that returns this unnecessary slice

File: cmd/ignite-spawn/ignite-spawn.go [hunk 2]
```diff
@@ -56,11 +57,6 @@ func StartVM(vm *api.VM) error {
 	metricsSocket := path.Join(vm.ObjectPath(), constants.PROMETHEUS_SOCKET)
 	serveMetrics(metricsSocket)
```
Comment (luxas) [inline cmd/ignite-spawn/ignite-spawn.go]: Maybe use a waitgroup or something here to actually unblock the goroutine later when we know the firecracker process has started?

File: pkg/operations/start.go [hunk 3]
```diff
@@ -132,12 +132,19 @@ func StartVM(vm *api.VM, debug bool) error {
 		return err
 	}
 
-	// TODO: Enable this when ignite-spawn doesn't write to the file anymore
-	//if err := providers.Client.VMs().Set(vm); err != nil {
-	//	return err
-	//}
+	result, err := providers.Runtime.InspectContainer(containerID)
```
Comment (luxas) [inline pkg/operations/start.go]: You might want to add a comment above that we need to do this differently for CNI; there the IP address should be got from the CNI netns. We can follow that up

File: cmd/ignite-spawn/ignite-spawn.go [hunk 4]
```diff
@@ -80,22 +76,14 @@ func StartVM(vm *api.VM) error {
...
-func patchRunning(vm *api.VM, ipAddrs []net.IP) error {
-	return patchVM(vm, func(patchVM *api.VM) error {
-		patchVM.Status.Running = true
-		patchVM.Status.IPAddresses = ipAddrs
-		return nil
-	})
-}
-
 func patchStopped(vm *api.VM) error {
 	return patchVM(vm, func(patchVM *api.VM) error {
```
Comment (luxas) [inline cmd/ignite-spawn/ignite-spawn.go]: Now that the patch is only done in one place, and all the desired values are predictable, you can apply a static patch that you've defined in a string

File: cmd/ignite-spawn/ignite-spawn.go (hunk 2 above)
Comment (twelho) [reply cmd/ignite-spawn/ignite-spawn.go]: We can only get close to the startup place, as the Firecracker call is blocking. `ignite-spawn` cannot actually track Firecracker's state (without listening to some sockets), so this cannot be done at the moment.

File: pkg/operations/start.go (hunk 3 above)
Comment (twelho) [reply pkg/operations/start.go]: I'll add that

File: cmd/ignite-spawn/ignite-spawn.go (hunk 4 above)
Comment (twelho) [reply cmd/ignite-spawn/ignite-spawn.go]: Done in 872516a.

File: pkg/operations/start.go (hunk 3 above)
Comment (twelho) [reply pkg/operations/start.go]: Added in 94f7149.

File: cmd/ignite-spawn/ignite-spawn.go (hunk 1 above)
Comment (twelho) [reply cmd/ignite-spawn/ignite-spawn.go]: Done in 872516a.

**Comment by twelho** [comment]
The issue with `ignite-spawn`'s state clearing is solved with 872516a, the patcher was using metadata from when `ignite-spawn` started up as its base, which had been read while under update by Ignite's `StartVM` resulting in a race condition. That has now been prevented by using a static patch.

File: cmd/ignite-spawn/ignite-spawn.go (hunk 2 above)
Comment (luxas) [reply cmd/ignite-spawn/ignite-spawn.go]: Ok, good for now

**Comment by twelho** [comment]
> LGTM, thanks! Please create issues with the TODO followups from this PR

Done: https://github.com/weaveworks/ignite/issues/302 and https://github.com/weaveworks/ignite/issues/303

**Pull request merged**